backend/pkg/modules/internal/fuzzer/tools/specs.go

```diff
@@ -1,36 +1,163 @@
+type FuzzerSpecsInfo struct {
+	ProvidedSpec          string
+	ReconstructedSpec     string
+	ProvidedSpecInfo      *models.SpecInfo
+	ReconstructedSpecInfo *models.SpecInfo
+}
+
+func GetAPISpecsInfo(ctx context.Context, accessor core.BackendAccessor, apiID uint) (*FuzzerSpecsInfo, error) {
+	fuzzerSpecsInfo := FuzzerSpecsInfo{}
+
 	apiInfo, err := accessor.GetAPIInfo(ctx, apiID)
 	if err != nil {
 		return nil, fmt.Errorf("can't access to apiID (%v), error=%v", apiID, err)
 	}
 
-	specsInfo := &models.OpenAPISpecs{}
 	if apiInfo.ProvidedSpecInfo != "" {
 		specInfo := models.SpecInfo{}
 		if err := json.Unmarshal([]byte(apiInfo.ProvidedSpecInfo), &specInfo); err != nil {
 			return nil, fmt.Errorf("failed to unmarshal provided spec info. info=%+v: %v", apiInfo.ProvidedSpecInfo, err)
 		}
-		specsInfo.ProvidedSpec = &specInfo
+		fuzzerSpecsInfo.ProvidedSpecInfo = &specInfo
+	}
+
+	if apiInfo.ProvidedSpec != "" {
+		fuzzerSpecsInfo.ProvidedSpec = apiInfo.ProvidedSpec
 	}
 
 	if apiInfo.ReconstructedSpecInfo != "" {
 		specInfo := models.SpecInfo{}
 		if err := json.Unmarshal([]byte(apiInfo.ReconstructedSpecInfo), &specInfo); err != nil {
 			return nil, fmt.Errorf("failed to unmarshal reconstructed spec info. info=%+v: %v", apiInfo.ReconstructedSpecInfo, err)
 		}
-		specsInfo.ReconstructedSpec = &specInfo
+		fuzzerSpecsInfo.ReconstructedSpecInfo = &specInfo
+	}
+
+	if apiInfo.ReconstructedSpec != "" {
+		fuzzerSpecsInfo.ReconstructedSpec = apiInfo.ReconstructedSpec
+	}
+
+	return &fuzzerSpecsInfo, nil
+}
+
+func IsV2Specification(data []byte) bool {
+	specV2OrV3 := struct {
+		Openapi *string `json:"openapi"`
+		Swagger *string `json:"swagger"`
+	}{}
+	if err := yaml.Unmarshal(data, &specV2OrV3); err != nil {
+		return false
+	}
+	return specV2OrV3.Swagger != nil
+}
+
+func IsV3Specification(data []byte) bool {
+	specV2OrV3 := struct {
+		Openapi *string `json:"openapi"`
+		Swagger *string `json:"swagger"`
+	}{}
+	if err := yaml.Unmarshal(data, &specV2OrV3); err != nil {
+		return false
+	}
+	return specV2OrV3.Openapi != nil
+}
+
+func LoadSpec(spec []byte) (*openapi3.T, error) {
+	if IsV2Specification(spec) {
+		logging.Logf("[Fuzzer] getDocFromSpec(): spec V2 identified")
+		loader := openapi3.NewLoader()
+		var docV2 openapi2.T
+		if err := json.Unmarshal(spec, &docV2); err != nil {
+			panic(err)
+		}
+		doc, err := openapi2conv.ToV3(&docV2)
+		if err != nil {
+			panic(err)
+		}
+		if err := doc.Validate(loader.Context); err != nil {
+			panic(err)
+		}
+		return doc, nil
+	} else if IsV3Specification(spec) {
+		logging.Logf("[Fuzzer] getDocFromSpec(): spec V3 identified")
+		loader := openapi3.NewLoader()
+		doc, err := loader.LoadFromData(spec)
+		if err != nil {
+			panic(err)
+		}
+		if err := doc.Validate(loader.Context); err != nil {
+			panic(err)
+		}
+		return doc, nil
 	}
+	return nil, fmt.Errorf("invalid spec")
+}
 
-	return specsInfo, nil
+func GetBasePathsFromServers(servers *openapi3.Servers) []string {
```

Comment:
I've implemented the same differently [here](https://github.com/openclarity/speculator/pull/20/files#diff-15c06e366d2e41835cf86a9108e014f70974fb443ac61910273273b1e855f7faR103) to support servers w/o host names and with formatting (examples can be found [here](https://swagger.io/docs/specification/api-host-and-base-path/) and in the [unit test](https://github.com/openclarity/speculator/pull/20/files#diff-93600c0940ef11d86724a8e7a2b8df94c11df899d005e6a2067bdd1fb0e9e0c2R204))
maybe we can use import the logic from the speculator once merged. WDYT?

Reply:
If I'm right, you can specify one or more base URLs on openapi specs on "servers" part, then we need to return an array of string, at least for Fuzzer needs.
For the implementation, sure your code works. Using http.NewRequest in place of "manual" parsing allow to manage all "strange use cases" we can encountered when a user fill a URL manually (and use less of code).
But it doesn't matter, if you prefer I can reuse your code. But I will have to change it to manage list of string, and then I will have to change the behavior in all ApiClarity codes that use your function...

Reply:
ok, I understand. does `http.NewRequest` handle templating, for example `{protocol}://api.example.com/api`?

Reply:
No, you're right. I was not aware about templating in servers entries. I tested it (in case of) but of course it doesn't work. 
I will fix this.